pkg/provisioners/kafka/controller/channel/reconcile.go

```diff
@@ -0,0 +1,147 @@
+*/
+
+package channel
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/eventing/pkg/provisioners/kafka/controller"
+)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Channel resource
+// with the current status of the resource.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.log.Info("Reconciling channel", "request", request)
+	channel := &v1alpha1.Channel{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, channel)
+
+	if errors.IsNotFound(err) {
+		r.log.Info("could not find channel", "request", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		r.log.Error(err, "could not fetch Channel", "request", request)
+		return reconcile.Result{}, err
+	}
+
+	original := channel.DeepCopy()
+
+	// Reconcile this copy of the Channel and then write back any status
+	// updates regardless of whether the reconcile error out.
+	err = r.reconcile(channel)
+	if !equality.Semantic.DeepEqual(original.Status, channel.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+		if _, err := r.updateStatus(channel); err != nil {
+			r.log.Info("Failed to update channel status", "error", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Requeue if the resource is not ready:
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(channel *v1alpha1.Channel) error {
+	// See if the channel has been deleted
+	accessor, err := meta.Accessor(channel)
+	if err != nil {
+		r.log.Info("failed to get metadata", "error", err)
+		return err
+	}
+	deletionTimestamp := accessor.GetDeletionTimestamp()
+	if deletionTimestamp != nil {
+		r.log.Info(fmt.Sprintf("DeletionTimestamp: %v", deletionTimestamp))
+		//TODO: Handle deletion
+		return nil
+	}
+
+	// Skip Channel as it is not targeting any provisioner
+	if channel.Spec.Provisioner == nil || channel.Spec.Provisioner.Ref == nil {
+		return nil
+	}
+
+	// Skip channel not managed by this provisioner
+	provisionerRef := channel.Spec.Provisioner.Ref
+	clusterProvisioner, err := r.getClusterProvisioner()
+	if err != nil {
+		return err
+	}
+	// TODO: Is there a better way to compare ref?
+	if provisionerRef.Name != clusterProvisioner.Name || provisionerRef.Namespace != clusterProvisioner.Namespace {
+		return nil
+	}
+
+	// The provisioner must be ready
+	if !clusterProvisioner.Status.IsReady() {
+		r.log.Info("provisioner is not ready", "provisioner", clusterProvisioner)
+		return nil
```

Comment:
Should this return an error, so that this will get reconciled again?

Reply:
Yup, thanks, missed that. Added a test :)